### ISSUE TITLE
Precompute much more data for elliptic curves over number fields

### ISSUE THREAD

**Issue by JohnCremona** [LMFDB/lmfdb#1267]
See #1199 .  Currently quite a lot of data displayed for al elliptic curve over number fields is not stored in the database, but computed on the fly.

Here's what happens when we make an ecnf web page:

the corresponding Sage elliptic curve object is created
its conductor is computed and factored (without the need to factor it we could of course construct the conductor from the db)
the discriminant ideal is computed and factored. This involves working out whether we have a global minimal model, of if not then at which primes it is non-minimal
the regulator of the generators is computed. (We have gens for rather few of the curves in the db and I need to be reminded to store the regulator of any points which we do have.
the local data data at all primes of bad reduction is computed from the Sage curve

My plan: add code to the ecnf upload scripts to add data fields into the database at upload time, so that the WebEllipticCurve objects (class ECNF) do not need to create Sage elliptic curves.

**Comment by JohnCremona** [comment]
Now that #1318 is done (under review at least) I'll start on this one.  After #1318 I will do things a little differently which will simplify the coding and the transition.
 - 0. Work out exactly which additional data fields to precompute and store.
 - 1. Add to the current upload script (which already does some of the precomputations) to do everything and store it.  
 - 2. Run this on all my input data files (~300,000 curves) while uploading them to a brand new collection, say nfcurves2.
 - 3. Add all the indices needed to nfcurves2.
 - 4. Change the code in the ecnf module to read from nfcurves2 instead of nfcurves and use the precomputed data to avoid all computation on the fly (except the plot).
 - 5. Test this new setup thoroughly.

After that two things have to happen ideally simultaneously: (a) rename nfcurves2 to nfcurves (and nfcurves to nfcurves_old say); and (b) change the name nfcurves2 to nfcurves in the code (which should be in exactly one place).  These need to be done with care since (a) has to happen in the cloud (after copying nfcurves2 there) and in Warwick at the same time as the code changes get pushed.

I'm sure we can work out how to do the last step.  I would prefer not to do what I have done over Q which is to have code which works both with the old smaller data and the new larger data for each curve, since that was rather tedious.  but not impossible.

**Comment by AndrewVSutherland** [comment]
@JohnCremona I think we can handle this, but why not let the code use nfcurves2 for a while; there is no reason we can't just copy nfcurves2 to the cloud and push the code to prod.  Once we are happy (and have backed up everything) we can simply replace nfcurves with a copy (not a rename) of nfcurves2, then change the code to use nfcurves, push it, and then delete nfcurves2.

We could do the same thing at the database level to change modularcurves2 to modularcurves (this is much bigger of course, but after our cleanup we have more space to work with).

**Comment by JohnCremona** [comment]
A very good suggestion!  Now for the real work (not for a few days though).

**Comment by AndrewVSutherland** [comment]
@JohnCremona No rush.  And I will review your current PR once I get the PR I am working on now submitted.  BTW, I don't know if you saw the discussion about wtimeout that I had with Edgar in another thread, but I just made a change to rewrite.py to increase wtimeout to 60 seconds, which I believe should prevent the BulkWriteErrors you were occasionally seeing.  But please let me know if you run into any problems with it.

**Comment by JohnCremona** [comment]
AFter working on this more today I realise that one reason it is taking a long time is that I am trying to do several different things in one go, which I now think I will do in at least two passes:
   1. Precompute all the additional data required for each cure and add it to the database.  (The code to do this was written a long time ago, it just has not been run on everything yet: a partial collection is in nfcurves2).
   2. Change the website code to use the additional data, getting it from the new temporary collection nfcurves2 at first.
   3. Re-sort and re-label all curves using the new fixed deterministic sorting of primes and of curves in an isogeny class. Keep additional labels where relevant, e.g. HMF labels for curves over totally real fields, so that even if these are not the same as the new level+isogeny class label, it is possible to display this other label, and to search on it so as to keep the links between curves and newforms working.
   4. Simplify models for elliptic curves over fields of degree >2.  This will be done at the same time as 3.  Both 3 and 4 will involve making a new collection, and some code changes.

This Issue is actually about 1+2, which was well under way when I got sidetracked by 3+4.  My current plan is to see if I can complete 1+2 this week, and go back to working on 3+4 in August.

**Comment by AndrewVSutherland** [comment]
Good idea, getting 1+2 done would let us address #1358 and they are really independent of 3+4.

**Comment by JohnCremona** [comment]
I expect to have a PR for this on Friday after finishing the new upload to collection nfcurves.new.  This deals with items 1 and 2 above.  After it, no elliptic curves are constructed at all, the only computations being some field and ideal operations (multiplication but not factorization).

**Comment by AndrewVSutherland** [comment]
@JohnCremona Excellent, this should kill #1358 (and speed things up across the board).

**Comment by JohnCremona** [comment]
Looking at the (currently growing) http://beta.lmfdb.org/api/elliptic_curves/nfcurves.new/ I can see some things I might have done differently such as a more compact package in one string for the a-invariants as I have for gens and torsion_gens, but (unlike the case for elliptic curves over Q) we will never be searching on this field.  However @AndrewVSutherland  if you think that would be better I can restart the process (after the minor recoding and testing that would take).

**Comment by AndrewVSutherland** [comment]
@JohnCremona I don't think it's a big deal (given that you don't plan to search on them).

**Comment by JohnCremona** [comment]
This was due to finish overnight but power cuts mean that it was not (and right now I do not even have access to the code branch needed).  I may or may not be able to finish it today, it all depends.

**Comment by JohnCremona** [comment]
Status report: 

   1. code all ready and in branch JohnCremona/ecnf-extra-data but no pull request until I have finished re-uploading all the data.
   2. re-upload currently 43% complete, should finish in a few hours
   3. indexing then needed (a few seconds using the make_indices() function)
   4. inventory already updated in my clone, will push when ncurves.new moved to nfcurves
   5. new nfcurves.rand needed after ncurves.new moved to nfcurves

Manual testing already welcome but the normal test will not all pass until all the upload is finished, and things will not be fast until that is done and the indices created,

**Comment by JohnCremona** [comment]
OK -- done.  The branch referred to above works with the new collection nfcurves.new on beta, which has indices and all tests pass.  I have already updated the inventory description of nfcurves (with list if indices in place) though the collection name is still nfcurves.new.

**Comment by AndrewVSutherland** [comment]
@JohnCremona Great!  Are you going to move nfcurves.new to nfcurves (I assume the current code will still work with the new version?).  Let me know when I should move the new nfcurves to the cloud (we should do this before we merge your new code to master).

**Comment by JohnCremona** [comment]
No.  New code only works with new data and vice versa.   If it looks good then next step is to copy nfcurves.new to cloud pending pushing code to beta and prod.    It was too much extra effort to have code which worked with both collections as I made many formatting changes  (improved consistency).

**Comment by AndrewVSutherland** [comment]
OK, I copied nfcurves.new to the cloud: http://www.lmfdb.org/api/elliptic_curves/nfcurves.new/, and I also create a random object index nfcurves.new.rand for it (and on the Warwick db).

**Comment by JohnCremona** [comment]
Thanks.  Sorry l forgot to create the .rand.    If it looks ok to you could you merge with the fix you did for #1873 before pushing?  The new code will now work on both beta and prod databases,  and we can do the renaming of collections after that.

**Comment by AndrewVSutherland** [comment]
@JohnCremona I just created PR #1876 from your ecnf-extra-data branch, and git is happy to merge it to master with no conflicts (so no need to integrate #1874 into your branch).  I will start testing it now.

**Comment by AndrewVSutherland** [comment]
@JohnCremona Still testing but some quick feedback: 

- very happy to see the unit ideal displayed as (1) rather than (1196131131a5+594671805a4−5090922650a3−2837103059a2+2931172827a+799149822) as on http://beta.lmfdb.org/EllipticCurve/6.6.371293.1/1.1/a/1.
- pages definitely load faster

The only problem I have found so far is that j-invariant searches don't work (e.g. specify base field 2.0.8.1 and j-invariant 0) because the format of the jinv attribute in the database has changed but it doesn't look like the query has been changed accordingly.  Twist links don't work for the same reason.

This should be easy to fix.  I know you are on holiday this week and I don't want to bug you with this -- I will fork your branch, fix it, submit a new PR, and continue testing from there.

**Comment by AndrewVSutherland** [comment]
Also need to fix labels of favorite curves, I will take care of that as well.

**Comment by AndrewVSutherland** [comment]
The new code and data are now live on www.lmfdb.org and seem to be working fine.  I just copied nfcurves.new to nfcurves on both the Warwick and cloud dbs.  I'm going to change the code to again read from nfcurves, and once it has been pushed to beta/prod I'll delete the .new versions and close this issue.

**Comment by JohnCremona** [comment]
Thanks.  Perfect.

**Comment by AndrewVSutherland** [comment]
Done.
